src/mongo/jobs/DiscoverOutdatedComposites.class.php

```diff
@@ -0,0 +1,296 @@
+        // which include specific CBDs to regenerate
+        $regenTasksIncludingNulls = array_map($getRegenTaskForMetadata, $compositeMetadata);
+        $regenTasks = array_filter($regenTasksIncludingNulls, 'isset');
+
+        // for collected regeneration tasks, run the update
+        // TODO: this will probably want to schedule updates, rather than running directly
+        $this->runRegenerationTasks($regenTasks);
+    }
+
+    /**
+     * @param array $data
+     * @param string|null $queueName
+     */
+    public function createJob(Array $data, $queueName=null)
+    {
+        if(!$queueName)
+        {
+            $queueName = Config::getDiscoverQueueName();
+        }
+        elseif(strpos($queueName, \Tripod\Mongo\Config::getDiscoverQueueName()) === false)
+        {
+            $queueName = \Tripod\Mongo\Config::getDiscoverQueueName() . '::' . $queueName;
+        }
+        $this->submitJob($queueName,get_class($this),$data);
+    }
+
+    /**
+     * Validate arguments in the job message
+     * @return array
+     */
+    protected function getMandatoryArgs()
+    {
+        return array(
+            self::TRIPOD_CONFIG_KEY,
+            self::STORE_NAME_KEY,
+            self::CURSOR_LIMIT
+        );
+    }
+
+    
+    public function runRegenerationTasks($regenTasks) {
+        foreach ($regenTasks as $regenTask) {
+            $specification = $regenTask->specification;
+            $compositeCollection = $regenTask->compositeCollection;
+            $compositeRegenFunction = $regenTask->compositeRegenFunction;
+            
+            foreach($regenTask->cbdDocuments as $cbdDoc) {
+                $compositeRegenFunction($specification, $compositeCollection, $cbdDoc);
+            }
+        }
+    }
+
+    /**
+     * Generates regeneration task data for any outdated documents in a composite collection,
+     * by scanning that composite collection one cursor-page at a time.
+     *
+     * @param CompositeMetadata $metadatum - the composite collection to search
+     * @return CompositeRegenTask|null - a CompositeRegenTask, if outdated documents are found; null otherwise
+    **/
+    public function getRegenTaskForMetadata($metadatum, $cursorLimit) {
+        $filterOutdated = $metadatum->getOutdatedQueryComponent();
+        if(!isset($filterOutdated)) {
+            return null;
+        } else {
+            $compositeCollection = $metadatum->compositeCollection;
+            $specification = $metadatum->specification;
+            $cbdCollection = $metadatum->cbdCollection;
+            $compositeRegenFunction = $metadatum->compositeRegenFunction;
+
+            $cbdIdFields = array(_ID_KEY.'.'._ID_RESOURCE => 1, _ID_KEY.'.'._ID_CONTEXT => 1);
+
+            // find one cursor page at a time for this composite
+            $outdatedComposites = 
+                $compositeCollection
+                    ->find($filterOutdated, $cbdIdFields)
+                    ->limit($cursorLimit);
```

Comment:
Can we guarantee that we don't get starvation on rare data?

We should definitely track and log the number of records that are out-of-date for a given composite, so we can observe behaviour.